morpheus/_lib/src/messages/multi_tensor.cpp

```diff
@@ -0,0 +1,121 @@
+                                       std::size_t mess_offset,
+                                       std::size_t mess_count,
+                                       std::shared_ptr<morpheus::TensorMemory> memory,
+                                       std::size_t offset,
+                                       std::size_t count) :
+  DerivedMultiMessage(meta, mess_offset, mess_count),
+  memory(std::move(memory)),
+  offset(offset),
+  count(count)
+{}
+
+const TensorObject MultiTensorMessage::get_tensor(const std::string &name) const
+{
+    return get_tensor_impl(name);
+}
+
+TensorObject MultiTensorMessage::get_tensor(const std::string &name)
+{
+    return get_tensor_impl(name);
+}
+
+TensorObject MultiTensorMessage::get_tensor_impl(const std::string &name) const
+{
+    CHECK(this->memory->has_tensor(name)) << "Cound not find tensor: " << name;
+
+    // check if we are getting the entire input
+    if (this->offset == 0 && this->count == this->memory->count)
+    {
+        return this->memory->tensors[name];
+    }
+
+    // TODO(MDD): This really needs to return the slice of the tensor
+    return this->memory->tensors[name].slice({static_cast<cudf::size_type>(this->offset), 0},
+                                             {static_cast<cudf::size_type>(this->offset + this->count), -1});
+}
+
+void MultiTensorMessage::set_tensor(const std::string &name, const TensorObject &value)
+{
+    // Get the input slice first
+    auto slice = this->get_tensor(name);
+
+    // Set the value to use assignment
+    slice = value;
+}
+
+void MultiTensorMessage::get_slice_impl(std::shared_ptr<MultiMessage> new_message,
+                                        std::size_t start,
+                                        std::size_t stop) const
+{
+    auto sliced_message = DCHECK_NOTNULL(std::dynamic_pointer_cast<MultiTensorMessage>(new_message));
+
+    sliced_message->offset = start;
+    sliced_message->count  = stop - start;
+
+    // If we have more inference rows than message rows, we need to use the seq_ids to figure out the slicing. This
+    // will be slow and should be avoided at all costs
+    if (this->count != this->mess_count && this->memory->has_tensor("seq_ids"))
+    {
+        auto seq_ids = this->get_tensor("seq_ids");
+
+        // Determine the new start and stop before passing onto the base
+        start = seq_ids.read_element<int32_t>({(TensorIndex)start, 0});
+        stop  = seq_ids.read_element<int32_t>({(TensorIndex)stop - 1, 0}) + 1;
+    }
+
+    // Pass onto the base
+    DerivedMultiMessage::get_slice_impl(new_message, start, stop);
+}
+
+void MultiTensorMessage::copy_ranges_impl(std::shared_ptr<MultiMessage> new_message,
+                                          const std::vector<std::pair<size_t, size_t>> &ranges,
+                                          size_t num_selected_rows) const
+{
+    auto copied_message = DCHECK_NOTNULL(std::dynamic_pointer_cast<MultiTensorMessage>(new_message));
```

Comment:
Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.